Set-Partitions for Experimental

pinguly opened the pull request:
An implementation of set-partitions as used for easy quantum groups, see for example:
- Section 1.2 & 1.3 of https://arxiv.org/abs/1312.3857
- Section 1 of https://arxiv.org/abs/1509.00988
- Section 2 of https://arxiv.org/abs/1609.02321

pinguly committed:
move code to experimental

fingolfin commented on experimental/SetPartitions/src/AbstractPartition.jl:
```diff
@@ -0,0 +1,18 @@
+
+function ⊗(p::T, q::T) where {T <: AbstractPartition}
+    tensor_product(p, q)
+end
```
Missing trailing newline

pinguly replied:
Are the missing newlines there? If not, what is the best way to display them? All the other comments so far should haven been resolved.

fingolfin commented on experimental/SetPartitions/src/ColoredPartition.jl:
```diff
@@ -0,0 +1,164 @@
+import Base.copy
+
+"""
+ColoredPartition
```
```suggestion
    ColoredPartition
```

fingolfin commented on experimental/SetPartitions/src/ColoredPartition.jl:
```diff
@@ -0,0 +1,164 @@
+struct ColoredPartition <: AbstractPartition
+    partition::Partition
+    color_upper_points::Array{Int64, 1}
+    color_lower_points::Array{Int64, 1}
```
```suggestion
    color_upper_points::Vector{Int64}
    color_lower_points::Vector{Int64}
```
This should be adjusted everywhere else, too

thofma replied:
Also `Int` instead of `Int64`?

fingolfin commented on experimental/SetPartitions/src/ColoredPartition.jl:
```diff
@@ -0,0 +1,164 @@
+
+
+"""
+tensor_product(p::ColoredPartition, q::ColoredPartition)
```
```suggestion
    tensor_product(p::ColoredPartition, q::ColoredPartition)
```
These lines should be indented by four spaces, which in Markdown indicates a code block.

fingolfin commented on experimental/SetPartitions/src/ColoredPartition.jl:
```diff
@@ -0,0 +1,164 @@
+"""
+tensor_product(p::ColoredPartition, q::ColoredPartition)
+
+This function applies on p tensor product with q (in O(n)).
```
```suggestion
This function applies on `p` tensor product with `q` (in O(n)).
```

And I'd drop the "(in O(n))" bit here, it is completely unclear what it means anyway (e.g. what is `n`; and is this meant to refer to the complexity? Why is it part of the function description?) If you really want to mention it, I'd add it in a separate paragraph or note or callout.

fingolfin commented on experimental/SetPartitions/src/ColoredPartition.jl:
```diff
@@ -0,0 +1,164 @@
+- `q`: Second input colored partition
+
+# Returns
+- `p` tensor product `q`
```
This way of explaining arguments and return types is highly discouraged except for cases were there really is no good alternative (say because there are so many or so complicated parameters).

Please read https://docs.julialang.org/en/v1/manual/documentation/#Writing-Documentation for details.

Here, one could write something like this:

> Return the tensor product of two colored partitions `p` and `q`.

fingolfin commented on experimental/SetPartitions/src/ColoredPartition.jl:
```diff
@@ -0,0 +1,164 @@
+
+    ret::Array = [copy(get_upper_points(p)), copy(get_lower_points(p)), copy(p.color_upper_points), copy(p.color_lower_points)]
+
+        if lr
```
The indentation here seems wrong

fingolfin commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,464 @@
+    end
+    
+    # while new were found apply on them unary tensor and composition
+    stop_whole::Bool = false
```
This way of writing code with a variable specifier is rather unidiomatic.
```suggestion
    stop_whole = false
```

fingolfin commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,464 @@
+        (stop_whole, all_partitions, already_u, all_partitions_by_size, all_partitions_by_size_top_bottom, trace) = do_unary(to_unary, all_partitions, stop_whole, already_u, max_length, all_partitions_by_size, all_partitions_by_size_top_bottom, trace, spatial_rotation)
+
+        # store pairs that are candidates to get tensor product
+        to_tens::Set{Array} = Set()
```
```suggestion
        to_tens = Set{Array}()
```
Ideally this would use a fully concrete type like `Set{Vector{Int}}`...

fingolfin commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,464 @@
+            end
+        end
+    end
+end
```
Add trailing newline

fingolfin commented on experimental/SetPartitions/src/GenerateCategory.jl:
````diff
@@ -0,0 +1,464 @@
+- list of all partitions size n constructed from partitions in p
+
+# Examples
+```julia-repl
````
Better us a `jldoctest`, then your example will automatically serve as a test for your code.

fingolfin commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,464 @@
+    @assert length(methods(spatial_rotation)[1].sig.parameters)-1 == 0 || (p isa Array{SpatialPartition})
+
+    # store all candidates found
+    all_partitions::Set{AbstractPartition} = Set()
```
```suggestion
    all_partitions = Set{AbstractPartition}()
```
and so on -- your current code is unidiomatic

fingolfin commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,464 @@
+    all_partitions_by_size::Dict{Int, Set{AbstractPartition}} = Dict()
+
+    # all candidates stored in tuple with a dict for top and bottom size
+    all_partitions_by_size_top_bottom::Array = [Dict(), Dict()]
```
The `Array` is a worse type specifier than what is automatically inferred.
```suggestion
    all_partitions_by_size_top_bottom = [Dict(), Dict()]
```

fingolfin commented on experimental/SetPartitions/README.md:
```diff
@@ -0,0 +1,4 @@
+Implementation of set-partitions as used for easy quantum groups, see for example:
+* Section 1.2 & 1.3 of https://arxiv.org/abs/1312.3857
+* Section 1 of https://arxiv.org/abs/1509.00988
+* Section 2 of https://arxiv.org/abs/1609.02321
```
Please follow the format set by `experimental/ExperimentalTemplate/README.md`. Also, hyperlinks in Markdown should be surrounded by `<` and `>`, i.e.
```suggestion
* Section 1.2 & 1.3 of <https://arxiv.org/abs/1312.3857>
* Section 1 of <https://arxiv.org/abs/1509.00988>
* Section 2 of <https://arxiv.org/abs/1609.02321>
```

fingolfin replied:
Also I think a little more explanation should be in here, in particular, that this is *not* what ordinary mathematicians would think of when talking about partitions

fingolfin commented on experimental/SetPartitions/src/AbstractPartition.jl:
```diff
@@ -0,0 +1,18 @@
+"""
+AbstractPartition
```
```suggestion
    AbstractPartition
```

fingolfin commented on experimental/SetPartitions/src/AbstractPartition.jl:
```diff
@@ -0,0 +1,18 @@
+"""
+AbstractPartition
+
+Abstract type of Partition for classical-, colored- and spatial Partitions.
```
Hmm, "Abstract type of Partition" doesn't parse as a valid english sentence for me.  Also, if intended to be used as a noun, it should be in lowercase; if it is meant to refer to a type it should be surrounded by backticks.

fingolfin reviewed (dismissed):
Thank you for your contribution!

I left some general comments that should be addressed in all similar places they apply.

thofma commented on experimental/SetPartitions/src/ColoredPartition.jl:
```diff
@@ -0,0 +1,164 @@
+        error("Partition has no bottom part.")
+    end
+
+    ret::Array = [copy(get_upper_points(p)), copy(get_lower_points(p)), copy(p.color_upper_points), copy(p.color_lower_points)]
```
```suggestion
    ret = (copy(get_upper_points(p)), copy(get_lower_points(p)), copy(p.color_upper_points), copy(p.color_lower_points))
```
I suggest to use a tuple.

thofma commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,254 @@
+
+function get_lower_points(p::ColoredPartition)
+    p.partition.lower_points
+end
```
Better names (better in the sense of consistency with the rest of Oscar) would be `upper_points` and `lower_points`.

pinguly committed:
improved readme

sebvz777 committed:
Merge branch 'oscar-system:master' into nf/set_partitions

sebvz777 committed:
minor style changes

sebvz777 committed:
Vectors to Tuple and Type annotations

pinguly committed:
fix tests and indentation

sebvz777 committed:
type annotations

sebvz777 committed:
fixed docs

sebvz777 committed:
more types if possible

pinguly committed:
fix type annotition and typo

benlorenz commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,460 @@
+"""
+function construct_category(p::Vector, n::Int, tracing::Bool = false, max_artificial::Int = 0, spatial_rotation::Function = nothing_function() = nothing)
+    
+    @assert length(methods(spatial_rotation)[1].sig.parameters)-1 == 0 || (p isa Vector{SpatialPartition})
```
see above

benlorenz commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,460 @@
+                    all_partitions_by_size_top_bottom = add_partition_to_composition_dict(
+                            all_partitions_by_size_top_bottom, a)
+                end
+            elseif length(methods(spatial_rotation)[1].sig.parameters)-1 == 1
```
Checking the internal fields of a julia Method object seems very fragile (and already the `[1]` access fails for julia 1.6).
Maybe just use `spatial_rotation::Union{Function,Nothing}=nothing` as function argument and check `spatial_rotation !== nothing` ?

If you really need to check some function argument types you may be able to use like `hasmethod(f, Tuple{SomeType})`.

pinguly committed:
math operators, fix jldoctest

pinguly committed:
improve docstrings

thofma commented:
Newlines look fine.

In OP, there are three arXiv preprints mentioned, but they are not cited in the documentation/code comments. Is this on purpose?

pinguly commented:
I replaced the three links with a reference to Sebastian's Bachelor's thesis in the README. In his thesis, you can also find the definitions of the different kinds of partitions and some information on the algorithms we implemented.

thofma commented:
Thanks for the explanation. The failing doctests (https://github.com/oscar-system/Oscar.jl/actions/runs/6893545019/job/18753368231?pr=2977) look genuine.

pinguly committed:
fix doctest

pinguly commented:
The doctests seem to work now.

fingolfin commented on experimental/SetPartitions/src/AbstractPartition.jl:
```diff
@@ -0,0 +1,26 @@
+"""
+abstract type AbstractPartition end
+
+function composition(p::T, q::T) where {T <: AbstractPartition}
```
Why not `compose`?

fingolfin commented on experimental/SetPartitions/README.md:
```diff
@@ -0,0 +1,31 @@
+In the future, we plan to implement:
+* linear combinations of partitions
+* examples 
+of tensor categories in the framework of [TensorCategories.jl](https://github.com/FabianMaeurer/TensorCategories.jl)
```
```suggestion
* examples of tensor categories in the framework of [TensorCategories.jl](https://github.com/FabianMaeurer/TensorCategories.jl)
```

fingolfin commented on experimental/SetPartitions/README.md:
```diff
@@ -0,0 +1,31 @@
+
+## References
+* <a id="1">[1]</a>
+Teodor Banica and Roland Speicher. Liberation of orthogonal Lie groups.
```
Please use the `.bib` file and the `[BLA](@cite)` syntax provided by DocumenterCitations (just grep our code base for `@cite` to find examples.

fieker commented on experimental/SetPartitions/src/ColoredPartition.jl:
```diff
@@ -0,0 +1,141 @@
+"""
+    tensor_product(p::ColoredPartition, q::ColoredPartition)
+
+Return the tensor product of `p` and `q`.
```
is it obvious what a tensor prduct of partitions is? or can we get a reference or definition?

fieker commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,498 @@
+    construct_category(p::Vector{AbstractPartition}, n::Int, tracing::Bool = false, 
+        max_artifical::Int = 0, spatial_rotation::Union{Function,Nothing}=nothing)
+
+Return a list of all partitions of size `n` constructed from partitions in `p` without using partitions of size greater than max(`n`, maxsize(`p`), `max_artifical`)
```
To what extend is this a category?

fieker commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,498 @@
+"""
+    get_trace(trace::Dict, start::Partition)
+
+Print the trace of the partition `start` constructed with `construct_category`.
```
trace is any dictionary? Usually I'd expect the function to be called just `trace` as the "get" is automatically implied. But I don't understand the 'Dict'

fieker commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,190 @@
+    p.dimension == q.dimension && is_composable(p.partition, q.partition)
+end
+
+function is_worth_composition(p::AbstractPartition, q::AbstractPartition, max_length::Int)
```
definition?

fieker commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,190 @@
+        end
+    end
+
+    return all(i -> i == 0, values(block_to_size))
```
'all(iszero, values(block_to_size))`

fieker commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,95 @@
+    p_return
+end
+
+function add_partition_to_dict(dict::Dict{Int, Set{AbstractPartition}}, p::AbstractPartition)
```
why the `to_dict` part? Isn't that obvious from the arguments?
Maybe the 'Dict{Int, Set...}` should actually be a data type linked to the partitions? I'd assume if I cann this with a random Dict of this type, chaos will reign

fieker commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,95 @@
+    return dict
+end
+
+function add_partition_to_composition_dict(vector::Vector, p::AbstractPartition)
```
at leat in comment: what vector is this?

fieker commented on experimental/SetPartitions/src/SetPartitions.jl:
```diff
@@ -0,0 +1,65 @@
+export size
+export *
+export ⊗
+export ⋅
```
what is this for?

fingolfin commented on experimental/SetPartitions/src/ColoredPartition.jl:
```diff
@@ -0,0 +1,141 @@
+# Arguments
+- `p`: input partition
+- `lr`: whether left (true) or right (false)
+- `tb`: whether top (true) or bottom (false)
```
Reading this I still have no idea what `lr` and `tb` do :-(. In particular, "whether left or right" is not a complete sentence, and it doesn't say how the rotation is defined

fingolfin commented on experimental/SetPartitions/src/ColoredPartition.jl:
```diff
@@ -0,0 +1,141 @@
+function rotation(p::ColoredPartition, lr::Bool, tb::Bool)
+
+    if tb && isempty(upper_points(p))
+        error("Partition has no top part.")
```
No period at end of error messages like that

fingolfin commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,498 @@
+function do_unary(
```
This is a reaaally long argument list and complicated function -- it should have at least *some* comments. At the very, very least a line saying what is is about. Great would be a pointer to a textbook or paper that describes the algorithms. More of course would be better. Otherwise this becomes and unmaintainable blackbox down the road.

fingolfin commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,498 @@
+                            all_partitions_by_size_top_bottom, a)
+                end
+            elseif spatial_rotation !== nothing
+                a = spatial_rotation(pmod)
```
This part of the code is not covered by the tests; with such a complex function that makes me wary.

fingolfin commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,498 @@
+
+end
+
+function do_tensor_products(
```
Same comments as for `do_unary` apply here, too.

fingolfin commented on experimental/SetPartitions/src/GenerateCategory.jl:
```diff
@@ -0,0 +1,498 @@
+        all_partitions_by_size_top_bottom, trace)
+end
+
+function do_composition(
```
Same comments as for `do_unary` apply here, too.

fingolfin commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,190 @@
+            block_to_size[i] = 1
+        else
+            block_to_size[i] = get(block_to_size, i, -1) + 1
+        end
```
```suggestion
        block_to_size[i] = get(block_to_size, i, 0) + 1
```

fingolfin commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,190 @@
+        if !(n in already_seen)
+            push!(already_seen, n)
+            block_to_size[n] = get(block_to_size, n, -1) - 1
+            if (get(block_to_size, n, -1) > 0 && 
```
You literally just stored `block_to_size[n]`, why retrieve it again, and with `get`?
```suggestion
            block_to_size[n] = get(block_to_size, n, -1) - 1
            if (block_to_size[n] > 0 && 
```
or even better, avoid another hash table lookup:
```suggestion
            bs = block_to_size[n] = get(block_to_size, n, -1) - 1
            if (bs > 0 && 
```

fingolfin commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,190 @@
+        else
+            if p_vector[i-1] != n && get(block_to_size, p_vector[i-1], -1) != 0
+                return false
+            else
```
Since you `return` in the line above, you can just `end` here, and then de-indent the following block. Advantage: the code is nested less, which makes it either to read and understand.
```suggestion
            end
```

fingolfin commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,266 @@
+"""
+    SetPartition
+
+Initialize SetPartition object, while transforming object to "normal form". 
```
This "explanation" does not explain the most crucial thing: what does `SetPartition` represent?

pinguly replied:
Should we add a high-level explanation with a reference to a paper or do you prefer a more precise mathematical definition in the docstrings?

fingolfin commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,266 @@
+"""
+    tensor_product(p::SetPartition, q::SetPartition)
+
+Return tensor product of `p` and `q`.
```
Again, what is the definition of the tensor product? Should either be part of the docstring, or the docstring should state where the definition can be found

fingolfin commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,266 @@
+"""
+    involution(p::SetPartition)
+
+Return involution of `p`.
```
What is "the involution" of a set partition? Again, definition should be included, or a pointer to it.

fingolfin commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,266 @@
+        path = [ii]
+        already_in = Set()
+        push!(already_in, get(new_ids, ii, -1))
+        z = get(new_ids, ii, -1)
```
You get `z` here -- and then in the next line you overwrite it?!

fingolfin commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,266 @@
+            if z in already_in
+                break
+            end
+        end
```
Accessing entries of a dictionary is comparatively expensive. Don't do it again and again when you can get the data once and then keep it around. 

Also, if you create a `Set` via `Set()` this gives a `Set{Any}` which is not type stable, as it has to deal with *arbitrary* content. But here yo already know the type of the content. So you can write `Set{SomeType}()`. Even better: just create the set directly with the initial value, then Julia can infer the correct type.

So, here i a much more idiomatic way to write this code would be like this 
```suggestion
    for (ii, z) in new_ids
        path = [ii]
        already_in = Set((z,))
        while z in keys(new_ids)
            push!(path, z)
            push!(already_in, z)
            z = get(new_ids, z, -1)
            z in already_in && break
        end
```

fingolfin commented on experimental/SetPartitions/src/SpatialPartition.jl:
```diff
@@ -0,0 +1,70 @@
+"""
+    SpatialPartition
+
+Initialize SpatialPartition object
```
What is that?

fingolfin commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,95 @@
+Return a semantically identical partition in form of an vector, 
+which has new number values.
+"""
+function helper_new_point_values_vector(p::Vector, q::Vector)
```
Function decl does not match the docstring, why?
```suggestion
function helper_new_point_values_vector (p::Vector{Vector{Int}}, q::Vector{Vector{Int}})
```

fingolfin commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,95 @@
+"""
+function helper_new_point_values_vector(p::Vector, q::Vector)
+
+    p_points = vcat(copy(p[1]), copy(p[2]))
```
Redundant copy, since you `vcat` anyway?
```suggestion
    p_points = vcat(p[1], p[2])
```

fingolfin commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,95 @@
+    end
+    
+    q_points = vcat(copy(q[1]), copy(q[2]))
+    new_ids = Dict()
```
Not type stable. Perhaps like this (I am guessing the types
```suggestion
    new_ids = Dict{Int,Int}()
```

fingolfin commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,95 @@
+    upper = []
+    for (i, n) in enumerate(q[1])
+        push!(upper, get(new_ids, n, -1))
+    end
```
You don't need `i`. And `upper` is a `Vector{Any}`. How about
```suggestion
    upper = [get(new_ids, n, -1) for n in q[1]]
```
or perhaps even this (I did not try to reason whether `new_ids[n]` will be defined in all cases):
```suggestion
    upper = [new_ids[n] for n in q[1]]
```

fingolfin commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,95 @@
+    
+    lower = []
+    for (i, n) in enumerate(q[2])
+        push!(lower, get(new_ids, n, -1))
```
same comments are for `upper` apply

fingolfin commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,95 @@
+    end
+    
+    [upper, lower]
+    
```
```suggestion
    return [upper, lower]
```

fingolfin commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,95 @@
+            p_return[1][i] = new_id
+            new_id += 1
+        else
+            p_return[1][i] = get(new_ids, p_return[1][i], -1)
```
Huh, isn't this the sameas
```suggestion
            p_return[1][i] = get(new_ids, n, -1)
```

fingolfin commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,95 @@
+            p_return[2][i] = new_id
+            new_id += 1
+        else
+            p_return[2][i] = get(new_ids, p_return[2][i], -1)
```
same as above

fingolfin commented on experimental/SetPartitions/test/runtests.jl:
```diff
@@ -0,0 +1,6 @@
+using Test
+using Oscar
+
+include("Partition-test.jl")
```
I think this `runtests.jl` file can be deleted, as we automatically include all *.jl files inside `test` subdirectories?

fingolfin reviewed (commented):
This code is for experimental, and as such it doesn't have to be "perfect".

That said, at the very least the objects and operations provided by this code should be clear, which to me they currently aren't at all. It is also not clear to me what the intended applications of some of these things are -- who wants to use these "spatial partitions" (whatever they are) for what?

Beyond this, there is a lot more in here that should be improved before it could ever have a chance of getting out of experimental. But it could be done after a merge, if you promise to do so.


I did this review because I was asked to, but I'd really appreciate e if others also did this kind of review which just looks for general issues with the code. I am not particularly interested in set partitions myself.

sebvz777 committed:
implementation of feedback except of doc and types

pinguly committed:
fix references

pinguly committed:
improving some docstrings

lgoettgens commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,193 @@
+end
+
+function is_composable(p::SetPartition, q::SetPartition)
+    length(p.upper_points) == length(q.lower_points)
```
```suggestion
    return length(p.upper_points) == length(q.lower_points)
```
All functions with the keyword `function` in front of should use `return`.

pinguly committed:
improving helper functions in util

sebvz777 committed:
added return keywords

pinguly committed:
improving typing in GenerateCategory.jl

lgoettgens commented:
Could you click on the little `resolve` button on all the comments that you already took care of? That makes it a bit easier to see what is still in the works.

pinguly committed:
improving imports/exports

lgoettgens commented on docs/oscar_references.bib:
```diff
@@ -270,6 +270,17 @@ @Article{BN07
 
+@Article{BS09,
+  author        = {Banica, Teodor and Speicher, Roland},
+  title         = {Liberation of orthogonal {L}ie groups},
```
```suggestion
  title         = {Liberation of orthogonal Lie groups},
```
Our Bibtex parser doesn't like these

lgoettgens commented on docs/oscar_references.bib:
```diff
@@ -915,6 +935,15 @@ @Article{Gat96
+  author        = {Gromada, Daniel},
+  title         = {Compact matrix quantum groups and their representation categories},
+  year          = {2020},
+  url           = {https://publikationen.sulb.uni-saarland.de/bitstream/20.500.11880/29789/1/dis.pdf},
```
please provide a DOI instead of a url. `10.22028/D291-32389` should work in this case

lgoettgens commented on experimental/SetPartitions/README.md:
```diff
@@ -0,0 +1,20 @@
+## Aims
+
+The goal of this project is to
+bring set-partitions to OSCAR.
```
```suggestion
The goal of this project is to bring set-partitions to OSCAR.
```

lgoettgens commented on experimental/SetPartitions/README.md:
```diff
@@ -0,0 +1,20 @@
+
+In the future, we plan to implement:
+* linear combinations of partitions
+* examples of tensor categories in the framework of [TensorCategories.jl](https://github.com/FabianMaeurer/TensorCategories.jl)
```
Could you add some contact information about the people working and maintaining this? For an example. see experimental/FTheoryTools/README.md

lgoettgens commented on experimental/SetPartitions/src/AbstractPartition.jl:
```diff
@@ -0,0 +1,26 @@
+
+function ⋅(p::T, q::T) where {T <: AbstractPartition}
+    return composition(p, q)
+end
```
This symbol is used nowhere else in Oscar. And I think having * should be enough, right?

lgoettgens commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,293 @@
+    return p.lower_points == q.lower_points && p.upper_points == q.upper_points
+end
+
+function copy(p::SetPartition)
```
Instead of `copy`, you should implement `deepcopy_internal` (for all types)

lgoettgens commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,293 @@
+
+`SetPartition` stores two vectors `upper_points` and `lower_points`
+where same numbers correspond to points in the same subset. Both vectors are 
+automatically converted into normal form as returned by [`normal_form`](@ref).
```
This is about internals of the struct. Users shouldn't touch/see/modify them, so this information is useless to misleading in a docstring. If you want to give users a way to see/modify this stuff, you should write appropriate functions.

sebvz777 committed:
get(dict_name, key, -1) to dict_name[key]

lgoettgens commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,293 @@
+# Arguments
+- `p`: input partition
+- `lr`: rotating at the left (true) or at the right (false)
+- `tb`: rotating from top to bottom (true) or from bottom to top (false)
```
These are not present in the signature in Line 100, and are still not explained properly. Having two booleans as positional arguments seems the seed of future confusion for me, as one needs to remember their order. Maybe it would be better to have them as keyword arguments instead.

pinguly replied:
Would it be better to split it into four functions like `rotate_top_left`, `rotate_top_right`, ... and then document it separately?

lgoettgens commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,293 @@
+        error("SetPartition has no bottom part")
+    end
+
+    ret = (copy(p.upper_points), copy(p.lower_points))
```
you should use `deepcopy` whenever you want to create a proper copy.

lgoettgens commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,293 @@
+
+Return the composition of `p` and `q` as well as the number of removed loops.
+
+The composition of two partitions is obtained concatenating them horizontally
```
```suggestion
The composition of two partitions is obtained by concatenating them horizontally
```

lgoettgens commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,293 @@
+"""
+function composition_loops(p::SetPartition, q::SetPartition)
+    
+    !is_composable(p, q) ? error("format not fitting") : 
```
```suggestion
    @req is_composable(p, q) "Number of points mismatch" 
```
or similar. Please use `@req` instead of `error(...)` for input sanitation

sebvz777 replied:
Hi, thanks for your feedback! This @req macro does not seem to be part of the julia standard library. Do you mean by using @req that I should implement my own req macro or should I use the @req from DiscreteValueIteration (https://github.com/JuliaPOMDP/DiscreteValueIteration.jl) or am I missing sth?

lgoettgens replied:
This is part of Hecke and thus available in Oscar without any extra work. See https://docs.oscar-system.org/dev/Hecke/features/macros/#Miscellaneous for the docs.

lgoettgens replied:
But I agree that this should be added to the Oscar developer docs

sebvz777 replied:
ah I see, makes sense, thanks!

lgoettgens commented on experimental/SetPartitions/src/SetPartition.jl:
````diff
@@ -0,0 +1,293 @@
+
+# Examples
+```jldoctest
+julia> SetPartition([2, 4], [4, 99])
````
All constructible types should have a snake_case function to construct them. See the first bullet point in https://docs.oscar-system.org/dev/DeveloperDocumentation/styleguide/#Naming-conventions

lgoettgens commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,108 @@
+Return two vectors which are semantically identical to `p_upper` and `p_lower`
+as set-partition and are numbered from 1 to the number of subsets.
+"""
+function normal_form(p_upper::Vector{Int}, p_lower::Vector{Int})
```
Is this a function that a user needs to call themselves? If not, I would call it `_normal_form` instead to make it clear that it is internal. 
I have seen calls to it in basically all constructors and operations, thus the question above.

lgoettgens commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,194 @@
+function size(p::SetPartition)
+    return length(p.lower_points) + length(p.upper_points)
+end
```
I find all of these concrete implementations weird in this file and would personally look for them in the corresponding PartitionType.jl file instead.

lgoettgens commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,194 @@
+end
+
+function upper_points(p::ColoredPartition)
+    return p.partition.upper_points
```
```suggestion
    return upper_points(p.partition)
```
avoid direct access as much as possible

lgoettgens replied:
not just here, but basically everywhere. In the optimal case, you only have direct access in one function per field and everything else refers to that.

lgoettgens commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,194 @@
+    return length(p.upper_points) == length(q.lower_points)
+end
+
+function upper_points(p::SetPartition)
```
for my use-case, I would like to have accessors for the number of upper/lower points, e.g.
```julia
function num_upper_points(p::AbstractPartition)
	return length(upper_points(p))
end
```

lgoettgens commented on experimental/SetPartitions/src/PartitionProperties.jl:
```diff
@@ -0,0 +1,194 @@
+Return whether it is worth to compose `p` and `q` such that the result is less
+equal `max_length` and not equal to the empty partition.
+
+Note that this is a helper function for the `construct_category` algorithm.
```
If it's purely an internal helper, consider prepending the name by a `_`.

lgoettgens reviewed (commented):
Some more comments from a quick look over some parts. I will have a look later, once some things are adapted.
As Max, I don't know a lot about the topic, so most comments are just from a code style perspective.

pinguly committed:
move getters from PartitionProperties, add getters

pinguly committed:
fix readme line break, add contact

pinguly committed:
remove runtests and cdot operator

pinguly committed:
Update docs/oscar_references.bib

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

pinguly committed:
references url -> doi

sebvz777 committed:
changed: deepcopy, @req, _normal_form

lgoettgens commented on experimental/SetPartitions/src/SetPartition.jl:
```diff
@@ -0,0 +1,305 @@
+
+function deepcopy(p::SetPartition)
+    return SetPartition(deepcopy(p.upper_points), deepcopy(p.lower_points))
+end
```
Please not that `deepcopy` is indeed the method to call for a copy. When providing a method however, you need to define `deepcopy_internal` (see https://docs.julialang.org/en/v1/base/base/#Base.deepcopy).

sebvz777 committed:
added _ to helper functions

sebvz777 committed:
small style correction

pinguly committed:
deepcopy_internal

pinguly committed:
constructors and renaming

pinguly committed:
add more docstrings

sebvz777 committed:
added requirements micro to constructor

sebvz777 committed:
tests for construcotr requirements

sebvz777 committed:
style changes

sebvz777 committed:
requirements check in constructor without helper

sebvz777 committed:
make doctests great again

lgoettgens commented on experimental/SetPartitions/src/Util.jl:
```diff
@@ -0,0 +1,160 @@
+
+Note that this is a helper function for the constructor of colored partitions.
+"""
+function _req_colored_partition(partition::AbstractPartition, 
```
I, personally, wouldn't have these as a separate function and just inline them into the constructor. This makes it furthermore possible to use multiple `@req` statements with different error messages tho help a user with wrong input.

pinguly committed:
use getters everywhere

pinguly committed:
fix some formatting, docstrings and tests

sebvz777 committed:
splitting of rotate into rotate_top_left, etc

pinguly committed:
fix doctests

pinguly committed:
Merge branch 'oscar-system:master' into nf/set_partitions

pinguly commented:
Is there anything else we should do for this first pull request? We definitely improve the code further and add more tests and documentations in the future.

lgoettgens reviewed (approved):
I think we do not have anybody who can review this from a technical pov.
The things that were mentioned here have all been adapted. If the authors want to continue working on this, I see no reason why not to merge this into experimental right now.

lgoettgens reviewed (changes requested):
I have one minor change I would like to have in here before merge:
There is an overhaul of method naming currently in action for `numer_of_things` functions (see https://github.com/oscar-system/Oscar.jl/pull/3272). To not make the code more inconsistent again, I would ask you to rename `num_lower_points` to `number_of_lower_points` and similarly for upper.

pinguly committed:
rename num -> number_of

lgoettgens reviewed (approved):
Thanks!

I think we do not have anybody who can review this from a technical pov.
The things that were mentioned here have all been adapted. If the authors want to continue working on this, I see no reason why not to merge this into experimental right now.

thofma commented:
Will merge this once we got https://github.com/oscar-system/Oscar.jl/pull/3288/ merged.

lgoettgens commented:
> Will merge this once we got #3288 merged.

Bump
